department-of-veterans-affairs/caseflow
# MattT/APPEALS-7867-and-APPEALS-7873

Resolves the issues detailed in [APPEALS-7867](https://vajira.max.gov/browse/APPEALS-7867) and [APPEALS-7873](https://vajira.max.gov/browse/APPEALS-7873)

### Description
Whenever converting a scheduled hearing to virtual, as a hearings coordinator, hearings scheduler, or VSO employee (if representing the hearing's claimant), send all email and timezone values to the backend to be persisted (including pre-populated values).

### Acceptance Criteria
- [ ] Whenever converting a hearing to virtual as any type of user, if pre-populated values placed within the hearing conversion form are not removed, all remaining values are sent to the backend to be persisted.
- [ ] Whenever converting a hearing to a type other than hearing, the pre-existing behavior where only altered values are sent to the backend is left intact.
- [ ] Email notifications are sent to all parties registered to a hearing upon converting to/from virtual.

### Testing Plan

**Note**: I highly recommend that you read [this comment](https://github.com/department-of-veterans-affairs/caseflow/pull/17371#discussion_r944053872) prior to working through the testing plan as it details how to set up a local email server for observing whether or not email notifications are being sent properly. 

1. Locate appeals with `assigned` `ScheduleHearingTasks`:
```ruby
uhs = Appeal.all.filter { |a| a.tasks.any? { |t| t.type == ScheduleHearingTask.name && t.status == "assigned" } }
```

Be sure that the hearing request type is not `virtual`.

```ruby
non_virtual_uhs = uhs.reject { |h| h.changed_hearing_request_type == "R" || (!h.changed_hearing_request_type && h.original_hearing_request_type == "R") }
```

2. Log into Caseflow as a hearings coordinator (ex: `BVASYELLOW`)

3. Navigate to the case details page for one of the appeals you've located:
`/queue/appeals/<non_virtual_uhs[n].external_id>`

4. Select the "Schedule Veteran" option from the actions dropdown.

5. Schedule the hearing **without** entering email addresses for the appellant and representative. Also be sure to keep the hearing set to be non-virtual..

6. Refresh the Case Details page, and then scroll down to the Hearings section. Click on the "View Hearing Details" link.

7. Select "Virtual" from the "Hearing Type" dropdown.

8. Leave any pre-populated values intact. You can choose to leave the representative email address field blank.

9. If you'd like, open the Developer Tools panel (CTRL+SHIFT+I in Chrome), and open the Network tab. Submit the conversion for. You should have sent a payload similar to this:

**Note**: Represenative's email address was left blank.
```json
{
    "appellant_tz": "America/Los_Angeles",
    "appellant_email_address": "prepopulated@email.com",
    "representative_tz": "America/Los_Angeles",
    "transcription_attributes": {},
    "virtual_hearing_attributes": {
        "request_cancelled": false,
        "job_completed": false
    },
    "email_recipients_attributes": {
        "0": {
            "id": "",
            "timezone": "America/Los_Angeles",
            "email_address": "prepopulated@email.com",
            "type": "AppellantHearingEmailRecipient"
        }
    }
}
```

10. If you've elected to run a local mail server, now's the time to check your inbox! You should have received an email message addressed to the appellant and/or representative confirming that their virtual hearing has been scheduled:
<img width="942" alt="maildev-test" src="https://user-images.githubusercontent.com/99351305/184781443-9786bb87-4a3b-4218-a8cb-b02d1bb33cc0.PNG">


11. You are also able to check that any email messages have successfully been sent at the bottom of the "Email Notifications" section of the Hearing Details page:

<img width="379" alt="emails_sent" src="https://user-images.githubusercontent.com/99351305/184781354-a78d6f61-922c-44ed-8606-9b9aa00816b7.PNG">

12. Repeat these steps for another unscheduled hearing. However, once you've scheduled the hearing, switch to a VSO user (ex: `BILLIE_VSO`) and perform the conversion as that user instead.


### Code Documentation Updates
